File: ruby/luhn/luhn.rb
module Luhn
  def self.valid?(digit_string)
    # return false if digit_string.scan(/\d/).length < 2
    # return false if digit_string.scan(/(\d|\s)/).length != digit_string.length

    sum = digit_string
          .gsub(/\s/, '')
          .tap { |str| return false unless str[/\A\d\d+\z/] }
          .scan(/\d/)
          .reverse
          .each.with_index
          .reduce(0) do |total, (d, i)|
            modified = d.to_i > 4 ? (d.to_i * 2) % 10 + 1 : d.to_i * 2
            value = (i % 2).odd? ? modified : d.to_i
            total + value
          end

    (sum % 10).zero?
  end
end

File: typescript/luhn/luhn.ts
export function valid(digitString: string): boolean {
  if(digitString.match(/[^\d|\s]/g)) return false
  if(digitString.replace(/\s/g, '').length <= 1) return false

  var digitStrings = digitString.match(/\d/g)?.reverse()
  var total = digitStrings?.map(Number)
                           .map(toLuhn)
                           .reduce(sum)

  return Number(total) % 10 === 0
}

function toLuhn(digit: number, index: number): number {
    let tmp = index % 2 === 1 ? digit * 2 : digit
    return tmp > 9 ? tmp - 9 : tmp
}

function sum(num1: number, num2: number): number {
  return num1 + num2
}
